Infrastructure: add flatpak support

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Add flatpak support
#### Motivation for adding to Mudlet
AppImage alternative and possibly better Steam Deck support
#### Other info (issues closed, discussion etc)
Sound works!

vadi2 committed:
WIP

vadi2 committed:
Get most of dependencies in place - Mudlet can start building

vadi2 committed:
Add cleanup steps

vadi2 committed:
Comment out buggy rename-desktop

vadi2 committed:
Build completes OK

vadi2 committed:
Cleanup

vadi2 committed:
Add trailing newline

vadi2 commented on CI/org.mudlet.mudlet.yml:
```diff
@@ -0,0 +1,190 @@
+    sources:
+      - type: git
+        url: https://github.com/Mudlet/Mudlet.git
+        branch: development
```
I'm not yet clear on what the flatpak best practices here are.

SlySven replied:
Maybe we'd want a "Release" pack following official release versions and then a second that follows development each night that it changes - i.e. the same as our PTBs...

vadi2 replied:
I'm not sure what do you mean, rephase?

It doesn't seem practical to keep modifying the file after every release... although in hindsight we can, because we already have https://github.com/Mudlet/Mudlet/blob/development/.github/repo-metadata.yml which tracks this.

But it would not be practical for PTBs.

vadi2 commented:
Once you have the right tools installed, you can built with with:

```
cd CI
flatpak install flathub org.kde.Platform//5.15-23.08
flatpak install flathub org.kde.Sdk//5.15-23.08
flatpak-builder build-dir org.mudlet.mudlet.yml --force-clean --user --install
```

And run with:

```
flatpak run org.mudlet.mudlet
```

SlySven commented on CI/org.mudlet.mudlet.yml:
```diff
@@ -0,0 +1,190 @@
+  - --socket=x11
+  - --socket=wayland
+  - --device=dri
+  - --socket=pulseaudio
```
Do we *have* to use ~~pus~~*pulse*audio some people do not like to have Poetterring-werez on their machines?

vadi2 replied:
I suspect that's needed to get sound working. You don't have to use it on your machine, this is just asking flatpak to grant access to PA.

Xabre666 replied:
Just to point the obvious: Flatpak depends on xdg-desktop-portal, that is needed for integration of containerized apps with host desktop session. That portal depends on Pipewire (which provides  PA via pipewire-pulse). If you're someone who objects the use of a sound server such as Pulseaudio, now replaced with Pipewire, you wouldn't be using Flatpak in the first place, so having Mudlet Flatpak use PA is objectively a non issue, and should one wish to avoid it they'll need to stick to classic packaging formats.

SlySven replied:
*\<off-topic>Actually, my attitude to PA has softened a bit in the last few days as I have finally found out why it wasn't handling my main system's 7.1 sound card and treating it as a plain stereo device... - it is still Poettering-ware but I guess he is focussing on other things these days - given the lack of activity registered to him on their public gitlab server.\</off-topic>*

vadi2 committed:
Merge branch 'development' into add-flatpak

vadi2 commented:
I'll dust this off and resume working on this again.

vadi2 committed:
update boost to 1.83.0 and enable RelWithDebInfo build

vadi2 committed:
Fix PCRE build

vadi2 committed:
Remove invalid cmake param

vadi2 committed:
Add a sample flatpak workflow

vadi2 committed:
Use another action for CI job

vadi2 committed:
Fix path to flatpak .yml

vadi2 committed:
Add desktop file into flatpak

vadi2 committed:
Try and copy mudlet.desktop into applications

Commit:
Install mudlet.desktop to share/applications

Commit:
Remove unneeded post-install call for mudlet.desktop

Commit:
Fix location to mudlet.desktop

Commit:
Enable utf8 and unicode in PCRE

Commit:
Update base image to better match our runtime

vadi2 committed:
Merge branch 'development' into add-flatpak

Commit:
Align checkout call with original action

Commit:
Add verbose key

Commit:
Remove branch key - it does not seem to be used by other flatpak-builder users

Commit:
Make pcre8 config opts be a string instead of a list

Commit:
Convert config-opts back to a list

vadi2 commented:
I think the utf8 in regex patterns not working is fixed now - didn't test, but the `configure` output is looking positive.

Where I'm stuck at is getting the system shortcut in place - even though I updated `src/CMakeLists.txt` to install `mudlet.desktop`, this is not reflected in the CI build logs.

vadi2 committed:
Use local directory as source for flatpak, not Git

vadi2 commented:
> even though I updated src/CMakeLists.txt to install mudlet.desktop, this is not reflected in the CI build logs.

Fixed this by having flatpak look into the directory it's working with instead of downloading Mudlet sources from our git. Credit to Patrick on the flathub matrix for help.

vadi2 committed:
Rename app-id to id and add rename-icon

vadi2 committed:
Add icons to cmake install

vadi2 committed:
Fix rename-icon call

vadi2 committed:
Default to Wayland

vadi2 committed:
Rename mudlet.desktop to org.mudlet.mudlet.desktop

Flatpak best practices

vadi2 committed:
Revert "Rename mudlet.desktop to org.mudlet.mudlet.desktop"

This reverts commit 201f70641ce47a12b17b65bcf0c44500a937897d.

vadi2 commented:
Flatpak recommends that org.mudlet.mudlet.desktop is used instead of mudlet.desktop - but doing that change caused things to break.

I'll pull the CI building out of this PR into another PR to separate concerns. Meanwhile, this PR can be tested by downloading https://github.com/Mudlet/Mudlet/actions/runs/7560712405/artifacts/1176225196 and running `flatpak install mudlet.flatpak`.

Xabre666 commented:
Seems there's something wrong with edbee-lib, it fails for me when it comes to it during the build
`fatal: unable to access 'https://github.com/Mudlet/edbee-lib.git/': Could
  not resolve host: github.com`
`fatal: clone of 'https://github.com/Mudlet/edbee-lib.git' into submodule
  path '/run/build/mudlet/3rdparty/edbee-lib' failed`
Prior submodules are cloned without issue.

vadi2 commented:
I suspect that is either a connectivity issue on your end or a temporary outage on github. Try again.

Xabre666 commented:
> I suspect that is either a connectivity issue on your end or a temporary outage on github. Try again.

As said, prior modules are cloned without issue, network is working fine, repo exists and can be cloned otherwise (outside the build process), and it all used to work normally just 3 days ago.

EDIT: just cloned all submodules manually to 3rd-party and build continues normally.
It's just cloning during the flatpak-builder process that does not work. Maybe because it tries to clone into run dir  instead of 3rd-party?

vadi2 commented:
If I understand you correctly, you tried to run the builder without cloning the repo fully?

vadi2 commented:
I suspect this is the conversation we had back in 2019 with https://discord.com/channels/283581582550237184/283582439002210305/655561524424736778, but as we see now the flatpak that is produced does work and flatpak developers reviewed it too.

What makes you think that the repo and submodules shouldn't be cloned prior to the flatpak builder running? 

If we have to 'package' our dependencies separately, then ignoring sandboxing features, this is much like apt-get again...

Xabre666 commented:
> If I understand you correctly, you tried to run the builder without cloning the repo fully?

`git clone https://github.com/Mudlet/Mudlet.git -b add-flatpak`
then cd into CI folder and run
`flatpak-builder build-dir org.mudlet.mudlet.yml --force-clean --user --install`
Sdk and runtime already installed.
It configures and starts building subnmodules like it always did, qtkeychain, lcf, pugi,  then comes to edbee and fails. error message as above. It tries to clone missing submodules but fails. Some don't and are cloned successfully, some do and build fails.
My guess is that problem is in it attempting to clone in '/run/build/mudlet/3rdparty/edbee-lib' during build, instead of the 3rd-party dir within the source. When I manually clone it there and restart the build, it builds and runs normally.
P.S. App icon works.

P.P.S what made me think that submodules will be downloaded automatically is the fact that it tries to download them automatically.

vadi2 commented:
I see - you need to run `git clone --recursive <rest of command>`, try it that way.

Xabre666 commented:
That'll of course work.

vadi2 committed:
Remove flatpak pipeline (to be added in another PR)

keneanung reviewed (approved):
Not too well versed about the flatpak definition file, but they look sensible to me.
